design/src/rfcs/rfc0016_flexible_checksum_support.md

```diff
@@ -0,0 +1,1442 @@
+                "x-amz-checksum-sha256:ZOyIygCyaOW6GjVnihtTFtIS9PNmskdyMlNKiuyjfzw=".len(),
+            ],
+        };
+        let mut aws_chunked_body = AwsChunkedBody::new(checksum_body, aws_chunked_body_options);
+
+        let mut output = SegmentedBuf::new();
+        while let Some(buf) = aws_chunked_body.data().await {
+            output.push(buf.unwrap());
+        }
+
+        let mut actual_output = String::new();
+        output
+            .reader()
+            .read_to_string(&mut actual_output)
+            .expect("Doesn't cause IO errors");
+
+        let expected_output = "B\r\nHello world\r\n0\r\nx-amz-checksum-sha256:ZOyIygCyaOW6GjVnihtTFtIS9PNmskdyMlNKiuyjfzw=\r\n\r\n";
+
+        // Verify data is complete and correctly encoded
+        assert_eq!(expected_output, actual_output);
+
+        assert!(
+            aws_chunked_body
+                .trailers()
+                .await
+                .expect("checksum generation was without error")
+                .is_none(),
+            "aws-chunked encoded bodies don't have normal HTTP trailers"
+        );
+    }
+
+    #[tokio::test]
+    async fn test_empty_aws_chunked_encoded_body() {
+        let sdk_body = SdkBody::from("");
+        let checksum_body = ChecksumBody::new(sdk_body, "sha256");
+        let aws_chunked_body_options = AwsChunkedBodyOptions {
+            stream_length: Some(0),
+            chunk_length: None,
+            trailer_lens: vec![
+                "x-amz-checksum-sha256:47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU=".len(),
+            ],
+        };
+        let mut aws_chunked_body = AwsChunkedBody::new(checksum_body, aws_chunked_body_options);
+
+        let mut output = SegmentedBuf::new();
+        while let Some(buf) = aws_chunked_body.data().await {
+            output.push(buf.unwrap());
+        }
+
+        let mut actual_output = String::new();
+        output
+            .reader()
+            .read_to_string(&mut actual_output)
+            .expect("Doesn't cause IO errors");
+
+        let expected_output =
+            "0\r\nx-amz-checksum-sha256:47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU=\r\n\r\n";
+
+        // Verify data is complete and correctly encoded
+        assert_eq!(expected_output, actual_output);
+
+        assert!(
+            aws_chunked_body
+                .trailers()
+                .await
+                .expect("checksum generation was without error")
+                .is_none(),
+            "aws-chunked encoded bodies don't have normal HTTP trailers"
+        );
+    }
+}
+```
+
+### Sigv4 Update
+
+When sending checksum-verified requests with a streaming body, we must update the usual signing process. Instead of signing the request based on the request's body, we must sign it with a magic string instead:
+
+```HTTP
+Authorization: <computed authorization header value using "STREAMING-UNSIGNED-PAYLOAD-TRAILER">
+x-amz-content-sha256: STREAMING-UNSIGNED-PAYLOAD-TRAILER
+```
```

Comment:
How does the signer know which algorithm header is used in order to create this magic string?

Reply:
It relies on pre-existing functionality, like how we currently send unsigned streaming requests.

Reply:
I'll update the language to reflect that we're setting this string to tell the signer that we have unsigned trailers following the body